core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+      this.history = Lists.newArrayList(base.history());
+      this.properties = Maps.newHashMap(base.properties());
+      this.changes = Lists.newArrayList(base.changes());
+    }
+
+    public Builder upgradeFormatVersion(int newFormatVersion) {
+      Preconditions.checkArgument(
+          newFormatVersion >= this.formatVersion,
+          "Cannot downgrade v%s view to v%s",
+          formatVersion,
+          newFormatVersion);
+
+      if (this.formatVersion == newFormatVersion) {
+        return this;
+      }
+
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder setLocation(String newLocation) {
+      if (null != this.location && this.location.equals(newLocation)) {
+        return this;
+      }
+
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder setCurrentVersionId(int newVersionId) {
+      if (this.currentVersionId == newVersionId) {
+        return this;
+      }
+
+      this.currentVersionId = newVersionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      return this;
+    }
+
+    public Builder setCurrentSchemaId(Integer newSchemaId) {
+      if (null != this.currentSchemaId && this.currentSchemaId.equals(newSchemaId)) {
+        return this;
+      }
+
+      this.currentSchemaId = newSchemaId;
+      this.changes.add(new MetadataUpdate.SetCurrentSchema(newSchemaId));
+      return this;
+    }
+
+    public Builder addSchema(Schema schema) {
+      this.schemas.add(schema);
+      this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      return this;
+    }
+
+    public Builder setSchemas(Iterable<Schema> schemasToAdd) {
+      for (Schema schema : schemasToAdd) {
+        this.schemas.add(schema);
+        this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      }
+
+      return this;
+    }
+
+    public Builder addVersion(ViewVersion version) {
+      this.versions.add(version);
+      this.changes.add(new MetadataUpdate.AddViewVersion(version));
+      this.history.add(
+          ImmutableViewHistoryEntry.builder()
+              .timestampMillis(version.timestampMillis())
+              .versionId(version.versionId())
+              .build());
+      return this;
+    }
+
+    public Builder setVersions(Iterable<ViewVersion> versionsToAdd) {
+      for (ViewVersion version : versionsToAdd) {
+        addVersion(version);
+      }
+
+      return this;
+    }
+
+    public Builder addProperty(String key, String value) {
+      this.properties.put(key, value);
+      this.changes.add(new MetadataUpdate.SetProperties(ImmutableMap.of(key, value)));
+      return this;
+    }
+
+    public Builder setProperties(Map<String, String> propertiesToAdd) {
```

Comment:
How about `newProperties` instead? Using the verb "add" is misleading here as well.

Reply:
done